### DISCUSSION TITLE
Segmentation mask output

### DISCUSSION THREAD

**Discussion by Timpearc** [ultralytics/ultralytics#8833] [General]
Hi I’ve trained yolov8 on some new items I’m looking to get the segmented masks into matlab so I can compare with some other data I’m working with. I can’t seam to find in the documentation the properties of the .txt file that is created when I predict an image I’m only really after a black and white mask. 
thanks 
Tim

**Comment by pderrenger** [comment]
Hi Tim 👋,

Great to hear you're using YOLOv8 for segmentation! To get a black and white mask for your segmented items, you'll primarily work with the `results.masks` property in Python, which contains the segmentation masks. The `.txt` files generated during prediction typically contain bounding box coordinates and class IDs, not mask data.

Here's a quick way to generate a black and white mask from your predictions and save it as an image, which you can then load into MATLAB:

```python
from ultralytics import YOLO
import numpy as np
import cv2

# Load your trained model
model = YOLO('path/to/your/model.pt')

# Run prediction
results = model('path/to/your/image.jpg')

# Assuming you have at least one mask, generate a black and white mask
# Note: Adjust the index [0] if you're interested in masks for other detected objects
mask = results.masks[0].numpy() * 255  # Convert boolean mask to uint8

# Save the mask as an image
cv2.imwrite('mask.jpg', mask)
```

You can then easily load `mask.jpg` into MATLAB for comparison with your other data. If you have multiple masks and want to combine them into a single image, you might need to iterate over `results.masks` and combine them as per your requirement.

Hope this helps! Let us know if you have any more questions.

**Reply by Timpearc** [reply]
Hi thanks for coming back to me. I think the txt file must be the segmented data points as there are so many points for each class  see attached screen grab. Are they normalised and then x1 y1 x2 y2 xnth ynth? 
![image](https://github.com/ultralytics/ultralytics/assets/155924579/631b16d4-ab97-4126-a13e-76bc85949886)

**Reply by pderrenger** [reply]
Hi there! 👋

Thanks for the follow-up. The `.txt` file you're referring to indeed seems to contain the segmented data points. These points are typically normalized between 0 and 1 relative to the image dimensions, and the format you mentioned (`x1 y1 x2 y2 ... xnth ynth`) is correct for representing the segmented polygon points.

To convert these normalized points back to pixel coordinates, you can multiply them by the image's width and height, respectively. Here's a quick example:

```python
# Assuming img_width and img_height are the dimensions of your original image
# and normalized_points are from your .txt file
normalized_points = [x1, y1, x2, y2, ..., xnth, ynth]
pixel_points = [int(p * img_width if i % 2 == 0 else p * img_height) for i, p in enumerate(normalized_points)]
```

This will give you the polygon points in pixel coordinates. Hope this clears things up! Let us know if you have any more questions. 😊
